neo4j/internal/pool/pool.go

```diff
@@ -453,35 +452,36 @@ func (p *Pool) Return(ctx context.Context, c idb.Connection) {
 }
 
 func (p *Pool) OnNeo4jError(ctx context.Context, connection idb.Connection, error *db.Neo4jError) error {
-	switch error.Code {
-	case "Neo.ClientError.Security.AuthorizationExpired":
+	if error.Code == "Neo.ClientError.Security.AuthorizationExpired" {
 		serverName := connection.ServerName()
 		p.serversMut.Lock()
 		defer p.serversMut.Unlock()
 		server := p.servers[serverName]
 		server.executeForAllConnections(func(c idb.Connection) {
 			c.ResetAuth()
 		})
-	case "Neo.ClientError.Security.TokenExpired":
+	}
+	if error.HasSecurityCode() {
 		manager, token := connection.GetCurrentAuth()
 		if manager != nil {
-			if err := manager.OnTokenExpired(ctx, token); err != nil {
+			handled, err := manager.HandleSecurityException(ctx, token, error)
+			if err != nil {
 				return err
 			}
-			if _, isStaticToken := manager.(auth.Token); !isStaticToken {
+			if handled {
 				error.MarkRetriable()
 			}
 		}
-	case "Neo.TransientError.General.DatabaseUnavailable":
+	}
+	if error.Code == "Neo.TransientError.General.DatabaseUnavailable" {
 		p.deactivate(ctx, connection.ServerName())
-	default:
-		if error.IsRetriableCluster() {
-			var database string
-			if dbSelector, ok := connection.(idb.DatabaseSelector); ok {
-				database = dbSelector.Database()
-			}
-			p.deactivateWriter(connection.ServerName(), database)
+	}
+	if error.IsRetriableCluster() {
+		var database string
+		if dbSelector, ok := connection.(idb.DatabaseSelector); ok {
+			database = dbSelector.Database()
 		}
+		p.deactivateWriter(connection.ServerName(), database)
```

Comment:
This should probably happen before calling the auth manager as it's infallible. De-facto are the code-paths mutually exclusive so it doesn't really matter. But that might change in the future if `error.IsRetriableCluster()` or `error.HasSecurityCode()` change their semantic.

Reply:
Makes sense, I've moved the DatabaseUnavailable and IsRetriableCluster checks before the HasSecurityCode check